### PR TITLE
Source map

### PR DESCRIPTION
**Pull request by sergey-shandar**
This PR has several major changes
1. A new class `ParsedJson`, which has
  - a typed JSON (JSON which is deserialized into specific type), 
  - a JToken (raw deserialized JSON which has line, position information)
  - a file name.
2. ObjectPath is a sequence of functions which can generate a name of next child property. Because old and new swagger documents may have different paths to the same item (e.g. parameter), we use a function instead of a constant.
3. ObjectPath functions can resolve '$ref'.

New output
```json
{
  "id": "1039",
  "code": "AddedOperation",
  "message": "The new version is adding an operation that was not found in the old version.",
  "jsonref-old": "old/added_path.json#/paths/~1api~1Operations/post",
  "jsonpath-old": null,
  "location-old": "",
  "jsonref-new": "new/added_path.json#/paths/~1api~1Operations/post",
  "jsonpath-new": "paths./api/Operations.post",
  "location-new": "new\\added_path.json:31:15",
  "type": "Info",
  "docurl": "https://github.com/Azure/openapi-diff/tree/master/docs/rules/1039.md",
  "mode": "Addition"
}
```

### CODE REVIEW

File: openapi-diff/src/core/OpenApiDiff.Core/Logging/FileObjectPath.cs [hunk 1]
```diff
@@ -1,31 +0,0 @@
...
-using System.Collections.Generic;
-using System.Linq;
-using YamlDotNet.RepresentationModel;
-
-namespace OpenApiDiff.Core.Logging
-{
-    /// <summary>
-    /// Represents a path into an object stored in a file.
-    /// </summary>
-    public class FileObjectPath
```
Comment (vladbarosan) [inline openapi-diff/src/core/OpenApiDiff.Core/Logging/FileObjectPath.cs]: I see the JsonReference and Readable path fields, is there any point in having ObjectPath public ?

Comment (vladbarosan) [inline openapi-diff/src/core/OpenApiDiff.Core/Logging/FileObjectPath.cs]: I see the JsonReference and Readable path fields, is there any point in having ObjectPath public ?

**Comment by vladbarosan** [comment]
@sergey-shandar can you describe your changes a little please ? that would be of great help for the reviewers

File: openapi-diff/src/modeler/AutoRest.Swagger/SwaggerParser.cs [hunk 2]
```diff
@@ -1,19 +1,26 @@
...
 using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using OpenApiDiff.Core;
 
 namespace AutoRest.Swagger
 {
     public static class SwaggerParser
     {
-        public static ServiceDefinition Parse(string swaggerDocument)
+        public static ParsedJson<ServiceDefinition> Parse(string swaggerDocument, string fileName)
```
Comment (veronicagg) [inline openapi-diff/src/modeler/AutoRest.Swagger/SwaggerParser.cs]: please add comments on what each param represents

File: openapi-diff/src/modeler/AutoRest.Swagger/Model/SwaggerParameter.cs [hunk 3]
```diff
@@ -42,9 +42,12 @@ public bool IsConstant
         /// <param name="context">The modified document context.</param>
         /// <param name="previous">The original document model.</param>
```
Comment (veronicagg) [inline openapi-diff/src/modeler/AutoRest.Swagger/Model/SwaggerParameter.cs]: please update comments

File: openapi-diff/src/modeler/AutoRest.Swagger/Model/ServiceDefinition.cs [hunk 4]
```diff
@@ -109,17 +109,21 @@ public ServiceDefinition()
...
+            }
+            if (context.PreviousRoot != previousDefinition)
+            {
+                throw new ArgumentException("context.PreviousRoot != previousDefinition");
+            }
 
-            var previousDefinition = previous as ServiceDefinition;
+            base.Compare(context, previousDefinition);
 
             if (previousDefinition == null)
```
Comment (veronicagg) [inline openapi-diff/src/modeler/AutoRest.Swagger/Model/ServiceDefinition.cs]: should this check move up?

File: openapi-diff/src/modeler/AutoRest.Swagger/Model/Response.cs [hunk 5]
```diff
@@ -66,19 +66,20 @@ public override IEnumerable<ComparisonMessage> Compare(ComparisonContext context
...
-                if (!headers.TryGetValue(header.Key, out newHeader))
+                if (!headers.TryGetValue(header.Key, out var newHeader))
                 {
                     context.LogBreakingChange(ComparisonMessages.RemovingHeader, header.Key);
                 }
 
                 context.Pop();
             }
 
+            context.PushProperty("schema");
```
Comment (veronicagg) [inline openapi-diff/src/modeler/AutoRest.Swagger/Model/Response.cs]: could you describe a bit the logic being used pushing properties to compare?

File: openapi-diff/src/modeler/AutoRest.Swagger/Model/Operation.cs [hunk 6]
```diff
@@ -143,21 +150,25 @@ public override IEnumerable<ComparisonMessage> Compare(ComparisonContext context
...
-                                             param : FindReferencedParameter(param.Reference, previousRoot.Parameters));
+                string.IsNullOrWhiteSpace(param.Reference) ? param : 
+                FindReferencedParameter(param.Reference, previousRoot.Parameters)
+            );
             foreach (var oldParam in priorOperationParameters)
             {
                 SwaggerParameter newParam = FindParameter(oldParam.Name, Parameters, currentRoot.Parameters);
 
-                context.PushProperty(oldParam.Name);
+                // context.PushProperty(oldParam.Name);
```
Comment (veronicagg) [inline openapi-diff/src/modeler/AutoRest.Swagger/Model/Operation.cs]: ??

File: openapi-diff/src/modeler/AutoRest.Swagger/ComparisonMessage.cs [hunk 7]
```diff
@@ -15,35 +19,56 @@ public class ComparisonMessage
     {
         public static string DocBaseUrl = "https://github.com/Azure/openapi-diff/tree/master/docs/rules/";
 
-        public ComparisonMessage(MessageTemplate template, FileObjectPath path, Category severity, params object[] formatArguments)
+        public ComparisonMessage(
+            MessageTemplate template,
+            ObjectPath path,
+            IParsedJson old,
+            IParsedJson @new,
```
Comment (veronicagg) [inline openapi-diff/src/modeler/AutoRest.Swagger/ComparisonMessage.cs]: why '@"?

File: openapi-diff/src/core/OpenApiDiff.Core/Logging/FileObjectPath.cs (hunk 1 above)
Comment (sergey-shandar) [reply openapi-diff/src/core/OpenApiDiff.Core/Logging/FileObjectPath.cs]: There is two modules: `OpenApiDiff.Core` and `AutoRest.Swagger`. `ObjectPath` is declared in the first and used in the second. But, I will make `ComparisonMessage.Path` as private.

File: openapi-diff/src/modeler/AutoRest.Swagger/ComparisonMessage.cs [hunk 8]
```diff
@@ -55,14 +80,26 @@ public ComparisonMessage(MessageTemplate template, FileObjectPath path, Category
         /// </summary>
         public MessageType Mode { get; }
 
+        private static string Location(IParsedJson j, JToken t)
```
Comment (veronicagg) [inline openapi-diff/src/modeler/AutoRest.Swagger/ComparisonMessage.cs]: could you add some comments here and describe format

File: openapi-diff/src/modeler/AutoRest.Swagger/ComparisonContext.cs [hunk 9]
```diff
@@ -1,38 +1,41 @@
...
             
             if (settings != null)
             {
-                this.Strict = settings.Strict;
+                Strict = settings.Strict;
             }
         }
 
         /// <summary>
         /// The original root object in the graph that is being compared
```
Comment (veronicagg) [inline openapi-diff/src/modeler/AutoRest.Swagger/ComparisonContext.cs]: please update comments

File: openapi-diff/src/modeler/AutoRest.Swagger/ComparisonContext.cs [hunk 10]
```diff
@@ -1,38 +1,41 @@
...
         /// </summary>
-        public object CurrentRoot { get; set; }
+        public T CurrentRoot => _CurrentRoot.Typed;
```
Comment (veronicagg) [inline openapi-diff/src/modeler/AutoRest.Swagger/ComparisonContext.cs]: what's "Typed" here?

File: openapi-diff/src/modeler/AutoRest.Swagger/ComparisonMessage.cs [hunk 11]
```diff
@@ -15,35 +19,56 @@ public class ComparisonMessage
...
 
         /// <summary>
         /// The JSON document path to the element being validated.
         /// </summary>
-        public FileObjectPath Path { get; }
+        public ObjectPath Path { get; }
+
+        public string OldJsonRef => Path.JsonPointer(Old);
+
+        public JToken OldJson() => Path.CompletePath(Old.Token).Last().token;
```
Comment (veronicagg) [inline openapi-diff/src/modeler/AutoRest.Swagger/ComparisonMessage.cs]: could you include somewhere a bit of info on JToken and how you're adding location to each element? thanks!

File: openapi-diff/src/modeler/AutoRest.Swagger/Model/Response.cs (hunk 5 above)
Comment (sergey-shandar) [reply openapi-diff/src/modeler/AutoRest.Swagger/Model/Response.cs]: it was a bug. the code creates an ObjectPath by pushing properties to the context. We have to push `schema` here.

File: openapi-diff/src/modeler/AutoRest.Swagger/ComparisonMessage.cs (hunk 7 above)
Comment (sergey-shandar) [reply openapi-diff/src/modeler/AutoRest.Swagger/ComparisonMessage.cs]: new is a keyword. We use `@` to say C# compiler that this is an identifier. Anyway, I renamed to `newDoc`

**Comment by veronicagg** [comment]
Updates look good to me, could you add a test to check on location updates you're making? thanks!

**Comment by sergey-shandar** [comment]
@veronicagg I will add it as soon as I get confirmation about format of location from @sarangag.

**Comment by sarangan12** [comment]
1. Code changes to add jsonPath to output
2. Future: Add length of the word with issue

**Pull request merged**